### ISSUE TITLE
Test hangs when using mp.Pool on Windows/macOS using CLI runner (PyCharm runner ok)

### ISSUE THREAD

**Issue by DylanLukes** [pytest-dev/pytest#10965]
This is a duplicate of #5429, which is _not_ resolved, but I come bearing more details, in particular a matrix that should help triage the source of the problem.

Here's a minimal example:

```python
from multiprocessing import Pool

def times_two(x):
    return x * 2

def test_foo():
    pool = Pool(4)
    assert pool.map(times_two, [1, 2, 3, 4]) == [2, 4, 6, 8]
```

Tests were performed on Github Actions `ubuntu-latest`, `windows-latest`, and `macos-latest`, which correspond to Ubuntu 22.04, Windows Server 2022, and macOS 12 Monterey ([not Ventura](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)). I also validated the macOS and Linux results on macOS 13 Ventura and Ubuntu 20.04 on my own machines. 

All of the below were tested with Python 3.11 on Github Actions runners, as well as 3.10 for those also run on my own machines.

| Platform | Runner | Result |
| ----------|---------|-------|
| Windows | `pytest` | Hangs |
| macOS | `pytest` | Hangs |
| Linux | `pytest` | Completes |
| macOS | PyCharm | Completes |

The last row in particular suggests that the issue is likely in the test runner. Running all tests via `PyCharm.app/Contents/plugins/python/helpers/pycharm/_jb_pytest_runner.py` does not hang.

– Dylan

**Comment by DylanLukes** [comment]
After some further troubleshooting, I realized I've missed a key element of reproduction:

```toml
[tool.pytest.ini_options]
addopts = [
    "--import-mode=importlib",
]
```

**Comment by evgeniiz321** [comment]
I faced the same problem on my mac. If run pytest with -s option there is a helpful stacktrace:

`Process SpawnPoolWorker-4:
Traceback (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/multiprocessing/queues.py", line 368, in get
    return _ForkingPickler.loads(res)
ModuleNotFoundError: No module named 'test_main'`

Looks like the problem is in the spawn method used by the multiprocessing by default on mac (and on windows). It doesn't propagate a list of modules known to a parent process to its subprocesses. importlib option doesn't affect sys.path as well due to its design. 

There are multiple workarounds here:

It helps to replace the default 'spawn' with the 'fork' (but only on mac), if use the example above:

`from multiprocessing import get_context


def times_two(x):
    return x * 2


def test_foo():
    with get_context("fork").Pool(1) as pool:
        assert pool.map(times_two, [1, 2, 3, 4]) == [2, 4, 6, 8]`
        
Also, if we run pytest like - python -m pytest, it prepends current directory to sys.path and in this case the scenario above also works. 

To sum up - I'm not sure this one can be fixed from the pytest perspective since it is how multiprocessing works on different platforms. 

Hope it helps someone :)

**Comment by RonnyPfannschmidt** [comment]
Importlib based imports are not compatible with multiprocess at all

**Comment by Jeitan** [comment]
Folks are still hitting this (#12178) - if these two things are incompatible, can we put that into the documentation?

**Comment by flying-sheep** [comment]
Let’s just fix it instead, this is so unexpected that before finding this issue today, if someone asked me “could that be the case”, I’d have confidently denied that this could be a possible issue – because why should it be?

**Comment by RonnyPfannschmidt** [comment]
fixing it involves making a import hook multiprocessing remotes to deal with the dissonance

**Comment by flying-sheep** [comment]
Sorry, I can’t grammatically parse this sentence (maybe there’s an “s” too much or a comma missing?), and that makes me think that your two comments are too terse to help us understand what’s going on.

What I understood so far: There seems to be some bug with the `importlib` mode that gets triggered by using multiprocessing that’s hard to fix. You seem to know the root cause of that bug, and it looks like you’re implying that fixing that root cause is hard.

**Comment by RonnyPfannschmidt** [comment]
we'd need to create a import hook for multiprocessing that allows importlib based modules to be sanely imported in the subprocess using the correct paths as importlib based imports arent working as normal imports

thats a very daunting task, at least im not going to put it on my todo

**Comment by flying-sheep** [comment]
I don’t get it:
1. `--import-mode=importlib` means that pytest uses `importlib` to import the test modules
2. users shouldn’t import these test modules in any way, instead reusability should be achieved by e.g. having fixtures in `conftest.py`

if one follows the rules in 2., there should be nothing that tries to “normally import” the test modules, right?

**Comment by RonnyPfannschmidt** [comment]
as per the traceback - multiprocessing not only needs to import them but also load their objects - that's not the user  - that's multiprocessing interacting with pickle

**Comment by flying-sheep** [comment]
OK, got it, so defining a function or class in a importlib-imported module and then passing it to something that pickles and (in another process) unpickles it doesn’t (necessarily?) work.

It probably works in some cases because of the `fork` method somehow preserving enough context, while `spawn` doesn’t. Your proposed fix is to add custom code (where?) that restores that context so unpicklers can find the correct file to grab the function/class object from.

Questions:
- where does the subtlety of the issue come from? Shouldn’t this fail more loudly? Is it not failing loudly a CPython bug that should be filed?
- how would hooking into multiprocessing work? Is that a feature or are you thinking on some hack (and maybe therefore not excited about the perspective)?

**Comment by RonnyPfannschmidt** [comment]
Creating a import hook that operates in worker processes and syncs things woul be something pytest has to do

There's numerous unknowns about Creating duch a hack and the complexity doesn't justify the wins
